### PR TITLE
Return non-200 HTTP response code if concurrent API requests cross the specified concurrent-request-limit #

### PR DESCRIPTION
**Pull request by spaparaju**
With Zero backlog and backlog getting timing out immediately with 1ms, all concurrent requests beyond 'middleware.concurrent-request-limit' result in immediate non-200 HTTP response

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -369,7 +369,9 @@ func main() {
 		r.Use(middleware.Recoverer)
 		r.Use(middleware.StripSlashes)
 		r.Use(middleware.Timeout(middlewareTimeout)) // best set per handler.
-		r.Use(middleware.Throttle(cfg.middleware.concurrentRequestLimit))
+		// With Zero backlog and backlog getting timing out immediately with 1ms, all concurrent requests beyond cfg.middleware.concurrentRequestLimit result in immediate non-200 HTTP response.
+		backlogDuration, _ := time.ParseDuration("1ms")
```
Comment (squat) [inline main.go:373]: Rather than parsing a string, which can return an error, let's use `time.Millisecond` directly

File: main.go [hunk 2]
```diff
@@ -369,7 +369,9 @@ func main() {
 		r.Use(middleware.Recoverer)
 		r.Use(middleware.StripSlashes)
 		r.Use(middleware.Timeout(middlewareTimeout)) // best set per handler.
-		r.Use(middleware.Throttle(cfg.middleware.concurrentRequestLimit))
+		// With zero backlog concurrent requests crossing a rate-limit result in non-200 HTTP response.
+		backlogDuration := 1 * time.Millisecond
```
Comment (kakkoyun) [inline main.go:373]: Let's make this configurable. We might need to increase this duration without building a new binary.

**Review by kakkoyun — CHANGES_REQUESTED** [review]
This change makes sense. We should be aware of the possible implications of having a large number of the backlog. As long as we have headroom resource-wise all would be good.

File: main.go [hunk 3]
```diff
@@ -707,6 +711,11 @@ func parseFlags() (config, error) {
 			" If not specified, local, non-shared rate limiting will be used.")
 	flag.IntVar(&cfg.middleware.concurrentRequestLimit, "middleware.concurrent-request-limit", 10_000,
 		"The limit that controls the number of concurrently processed requests across all tenants.")
+	flag.IntVar(&cfg.middleware.backLogLimitConcurrentRequests, "middleware.backlog-limit-concurrent-requests", 0,
+		"The number of concurrent requests that can buffered.")
+	flag.IntVar(&cfg.middleware.backLogDurationConcurrentRequests, "middleware.backlog-duration-concurrent-requests", 1,
```
Comment (kakkoyun) [inline main.go:716]:
```suggestion
	flag.DurationVar(&cfg.middleware.backLogDurationConcurrentRequests, "middleware.backlog-duration-concurrent-requests", time.Millisecond,
```

We have a duration type let's use it

**Review by kakkoyun — CHANGES_REQUESTED** [review]
Looking good. Only one thing.

**Review by kakkoyun — APPROVED** [review]

**Pull request merged**